### PR TITLE
Add instructions for creating change notes.

### PR DESCRIPTION
**Pull request by dbartol**

### CODE REVIEW

File: Adding change notes.md [hunk 1]
```diff
@@ -0,0 +1,69 @@
+# Adding change notes for query and library changes
+
+Each CodeQL query pack or library pack has its own change log to track how that pack changes with each release. Any non-trivial, user-visible change to a query or library should add a change note to the affected pack. This document describes how to do that.
+
+## Creating a change note
+To create a new change note for a pack, create a new markdown file in the `change-notes` directory of the pack (e.g., in `cpp/ql/src/change-notes` for the C++ standard query pack). The markdown file must be named `YYYY-MM-DD-id.md`, where `YYYY-MM-DD` is the date of the change, and `id` is a short string to help identify the change. Often, a good choice for the `id` portion of the filename is the name of the branch in which the changes were made. For example, if you were adding a new integer overflow query to the C++ standard query pack, you might do so from a branch named `int-overflow-query`, with a change note file named `cpp/ql/src/change-notes/2021-12-14-int-overflow-query.md`. Here is an example change note file:
```
Comment (adityasharad) [inline Adding change notes.md:6]: I think the branch name guidance may lead to some confusion, so perhaps drop that. The integer overflow example is good -- you can suggest `int-overflow-query` as the note ID without any connection to the branch.

File: Adding change notes.md [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+
+## How change notes affect the version of the pack
+When a new release of a pack is published, the pack's version number is advanced based on the categories of the changes in that version:
+- If any change note has a SemVer effect of `major version bump`, then the version number's major component will be incremented (e.g., `1.4.5` -> `2.0.0`).
+- Otherwise, if any change note has a SemVer effect of `minor version bump`, then the version number's minor component will be incremented (e.g., `1.4.5` -> `1.5.0`).
+- Otherwise, the version number's patch component will be incremented (e.g., `1.4.5` -> `1.4.6`).
+Thus, it is important to choose the correct category for each change note, so that users can rely on the pack's version number to indicate compatibility with previous versions of the pack.
+
+- 
+\e-notes` directory. The pack publishing process consolidates these individual change notes into the pack's `CHANGELOG.md` before publishing, removing the individual change note files in the process.
```
Comment (adityasharad) [inline Adding change notes.md:68]: Incomplete line?

File: Adding change notes.md [hunk 3]
```diff
@@ -0,0 +1,69 @@
+# Adding change notes for query and library changes
+
+Each CodeQL query pack or library pack has its own change log to track how that pack changes with each release. Any non-trivial, user-visible change to a query or library should add a change note to the affected pack. This document describes how to do that.
+
+## Creating a change note
+To create a new change note for a pack, create a new markdown file in the `change-notes` directory of the pack (e.g., in `cpp/ql/src/change-notes` for the C++ standard query pack). The markdown file must be named `YYYY-MM-DD-id.md`, where `YYYY-MM-DD` is the date of the change, and `id` is a short string to help identify the change. Often, a good choice for the `id` portion of the filename is the name of the branch in which the changes were made. For example, if you were adding a new integer overflow query to the C++ standard query pack, you might do so from a branch named `int-overflow-query`, with a change note file named `cpp/ql/src/change-notes/2021-12-14-int-overflow-query.md`. Here is an example change note file:
+
+```yaml
+---
+category: newQuery
```
Comment (adityasharad) [inline Adding change notes.md:10]: Examples are handy! Maybe add a couple more for common cases?

File: Adding change notes.md [hunk 4]
```diff
@@ -0,0 +1,69 @@
+# Adding change notes for query and library changes
```
Comment (adityasharad) [inline Adding change notes.md:1]: I suggest placing this file in https://github.com/github/codeql/tree/main/docs, and naming it something like `change-notes.md` or `change-notes-guide.md`.

File: Adding change notes.md [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+| feature        | minor version bump | A new library API has been added. |
+| majorAnalysis  | minor version bump | An API has changed in a way that may affect the results produced by a query that consumes the API. |
+| minorAnalysis  | patch version bump | An API has changed in a way that may affect the results produced by a query that consumes the API, but only in scenarios that affect relatively few users. |
+| fix            | patch version bump | An API has been fixed in a way that is not likely to affect the resultss produced by a query that consumes the API. |
+
+## How the final changelog is created
+When a new release of a pack is published, the publishing process consolidates all of the change note files from that pack's `change-notes` directory into the pack's `CHANGELOG.md` file. The consolidation process does the following:
+
+- Strips all metadata from each change note.
+- Strips all leading an trailing blank lines from the description of each change note.
```
Comment (adityasharad) [inline Adding change notes.md:54]:
```suggestion
- Strips all leading and trailing blank lines from the description of each change note.
```

File: Adding change notes.md (hunk 1 above)
Comment (dbartol) [reply Adding change notes.md]: Agreed. Fixed.

File: Adding change notes.md (hunk 2 above)
Comment (dbartol) [reply Adding change notes.md]: Oops. Removed.

File: Adding change notes.md [hunk 6]
```diff
@@ -0,0 +1,69 @@
+# Adding change notes for query and library changes
+
+Each CodeQL query pack or library pack has its own change log to track how that pack changes with each release. Any non-trivial, user-visible change to a query or library should add a change note to the affected pack. This document describes how to do that.
```
Comment (aeisenberg) [inline Adding change notes.md:3]: We should have a change note validation step as part of CI. It should just ensure the change notes are well-formed. Not blocking for this change, though.

Comment (dbartol) [reply Adding change notes.md:3]: Yeah, that's the plan. Probably just a switch to `codeql pack changelog` or something.

File: Adding change notes.md (hunk 3 above)
Comment (dbartol) [reply Adding change notes.md:10]: Added.

File: Adding change notes.md (hunk 4 above)
Comment (dbartol) [reply Adding change notes.md]: Moved to `docs/change-notes.md`, and added a link from `CONTRIBUTING.md`.

**Review by adityasharad — DISMISSED** [review]
Nice. I'm happy with this and we can keep improving it over time. Up to you if you want further review from the language teams, as they will be the ones writing these notes.

**Comment by dbartol** [comment]
Leaving this open for a day or so to allow for more comments.

File: docs/change-notes.md [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+
+### Library pack change categories
+| Category       | SemVer effect      | Description |
+|----------------|--------------------|-------------|
+| breaking       | major version bump | Any breaking change to the library pack, the most common of which is the deletion of an existing API. |
+| deprecated     | minor version bump | An existing API has been marked as `deprecated`. |
+| feature        | minor version bump | A new library API has been added. |
+| majorAnalysis  | minor version bump | An API has changed in a way that may affect the results produced by a query that consumes the API. |
+| minorAnalysis  | patch version bump | An API has changed in a way that may affect the results produced by a query that consumes the API, but only in scenarios that affect relatively few users. |
+| fix            | patch version bump | An API has been fixed in a way that is not likely to affect the resultss produced by a query that consumes the API. |
```
Comment (aschackmull) [inline docs/change-notes.md:69]:
```suggestion
| fix            | patch version bump | An API has been fixed in a way that is not likely to affect the results produced by a query that consumes the API. |
```

**Review by jbj — COMMENTED** [review]
Looks great!

**Review by aeisenberg — APPROVED** [review]
:shipit: 

Since this is a change to how users will contribute queries to this repository, it seems like we should add a change note somewhere about this change note. I don't think we need to be putting it into the change notes for the individual libraries, but possibly we can add a note to the CodeQL CLI with a pointer to these new instructions.

**Pull request merged**